src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -16,14 +16,235 @@
 
+#include "native_client/src/include/win/mman.h"
+#include "native_client/src/trusted/service_runtime/load_file.h"
+#include "native_client/src/trusted/service_runtime/sel_memory.h"
+
+#include "native_client/src/trusted/desc/nacl_desc_io.h"
+#include "native_client/src/shared/platform/lind_platform.h"
+#include "native_client/src/trusted/service_runtime/include/bits/mman.h"
+#include "native_client/src/trusted/service_runtime/include/sys/fcntl.h"
+
+
+/*
+ * always points at original program context
+ *
+ * -jp
+ */
+struct NaClThreadContext *master_ctx;
+
+
+/*
+ * dynamically allocate and initilize a copy
+ * of the parents NaClApp structure which is
+ * used in NaClSysFork()
+ */
+struct NaClApp *NaClChildNapCtor(struct NaClAppThread *natp) {
+  struct NaClApp *nap_child = NaClAlignedMalloc(sizeof *nap_child, __alignof(struct NaClApp));
+  struct NaClApp *nap_master = ((struct NaClAppThread *)master_ctx)->nap;
+  struct NaClApp *nap_parent = natp->nap;
+  NaClErrorCode *mod_status = NULL;
+
+  CHECK(nap_master);
+  CHECK(nap_parent);
+  CHECK(nap_child);
+
+  NaClLog(1, "%s\n", "Entered NaClChildNapCtor()");
+  memset(nap_child, 0, sizeof *nap_child);
+  if (!NaClAppCtor(nap_child)) {
+    NaClLog(LOG_FATAL, "%s\n", "Failed to initialize fork child nap");
+  }
+  mod_status = &nap_child->module_load_status;
+  nap_child->command_num = nap_parent->command_num;
+  nap_child->binary_path = nap_parent->binary_path;
+  nap_child->binary_command = nap_parent->binary_command;
+  nap_child->nacl_file = nap_parent->nacl_file ? nap_parent->nacl_file : LD_FILE;
+  nap_child->enable_exception_handling = nap_parent->enable_exception_handling;
+  nap_child->validator_stub_out_mode = nap_parent->validator_stub_out_mode;
+  nap_child->parent_id = nap_parent->cage_id;
+  nap_child->master = nap_master;
+  nap_child->parent = nap_parent;
+
+  /* fork generation count */
+  NaClXMutexLock(&nap_master->children_mu);
+  fork_num++;
+  NaClXMutexUnlock(&nap_master->children_mu);
+
+  /* store cage_ids in both master and parent to provide redundancy and avoid orphans */
+  for (struct NaClApp *nap_cur = nap_master; nap_cur; nap_cur = nap_parent) {
+    NaClXMutexLock(&nap_cur->children_mu);
+    if (nap_cur->children_ids[nap_cur->num_children]) {
+       nap_cur->num_children++;
+    }
+    /* make sure cage_id is unique and assign it to child */
+    InitializeCage(nap_child, nap_master->cage_id + fork_num);
+    NaClLog(1, "[nap %d] incrementing num_children\n", nap_cur->cage_id);
+    nap_cur->children_ids[nap_cur->num_children++] = nap_child->cage_id;
+    if (!DynArraySet(&nap_cur->children, nap_child->cage_id, nap_child)) {
+      NaClLog(LOG_FATAL, "[nap %u] failed to add cage_id %d\n", nap_cur->cage_id, nap_child->cage_id);
+    }
+    NaClLog(1, "[nap %d] new child count: %d\n", nap_cur->cage_id, nap_cur->num_children);
+    NaClXMutexUnlock(&nap_cur->children_mu);
+
+    /* break after parent's cleanup */
+    if (nap_cur == nap_parent) {
```

Comment:
I don't get the combination of this `if ... break` and `nap_cur = nap_parent` in the head of the `for` loop. Can the conditional `break` just be rolled into the termination condition of the `for` loop?

Reply:
This is just to perform the loop body operations on both parent and master. I think maybe it it possible squeeze the break condition into the for statement though, let me see if I can figure out a way.